llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+//===-- Cheerp/PartialExecuter.h - Analyze functions CFG to mark unreachable BBs --===//
+//
+//                     Cheerp: The C++ compiler for the Web
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright 2021-2022 Leaning Technologies
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "PartialExecuter"
+
+#include "../ExecutionEngine/Interpreter/Interpreter.h"
+#include "llvm/ADT/SCCIterator.h"
+#include "llvm/Cheerp/DeterministicUnorderedSet.h"
+#include "llvm/Cheerp/PartialExecuter.h"
+#include "llvm/ExecutionEngine/GenericValue.h"
+#include "llvm/IR/Operator.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/InitializePasses.h"
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+using namespace llvm;
+
+typedef cheerp::DeterministicUnorderedSet<const BasicBlock *, cheerp::RestrictionsLifted::NoErasure> DeterministicBBSet;
+
+namespace cheerp {
+
+char PartialExecuter::ID = 0;
+
+StringRef PartialExecuter::getPassName() const
+{
+	return "PartialExecuter";
+}
+
+PartialExecuter::PartialExecuter()
+	: llvm::ModulePass(ID)
+{
+}
+
+void PartialExecuter::getAnalysisUsage(AnalysisUsage& AU) const
+{
+}
+
+//Compute whether a given Value is a constant & it's value can be executed beforehand
+static bool isValueComputedConstant(const llvm::Value* V)
+{
+	// TODO(carlo): this can easily be memoized since there is no state
+	if (!isa<Constant>(V))
+		return false;
+	if (const GlobalVariable* GVar = dyn_cast<GlobalVariable>(V))
+	{
+		if (!GVar->hasInitializer() || GVar->isExternallyInitialized())
+			return false;
+		if (!GVar->hasInternalLinkage())
+			return false;
+	}
+	if (const ConstantExpr* CE = dyn_cast<ConstantExpr>(V))
+	{
+		for (auto& op : CE->operands())
+		{
+			if (!isValueComputedConstant(op))
+				return false;
+		}
+	}
+	return true;
+}
+
+class FunctionData;
+class ModuleData;
+
+class PartialInterpreter : public llvm::Interpreter {
+	friend FunctionData;
+	friend ModuleData;
+	const llvm::BasicBlock* fromBB{nullptr};	
+	enum BitMask : uint32_t
+	{
+		NONE = 0,
+		ALIGNED2 = 0x00000001,
+		ALIGNED4 = 0x00000003,
+		ALIGNED8 = 0x00000007,
+		ALL = 0xffffffff,
+	};
+	std::vector<std::pair<const llvm::Value*, const llvm::Value*> > incomingsVal;
+	std::vector<const llvm::PHINode*> notComputedPhis;
+	std::vector<std::pair<long long,long long>> validLoadIntervals;
+public:
+	GenericValue getOperandValueExplicit(const Value* V, ExecutionContext& EC)
```

Comment:
Is this function only used to apply a const-cast? Then it's better to not use constantness on our side

Reply:
Yes, this is to wrap the ExecutionEngine code that don't use const inside it's data structures (but values are const in some sense since we will not call any modifying instruction).
Removing (basically) all const our side also works.

Reply:
I'd say remove constness on our side where the engine does not agree. const_casts suggest that _we_ are messing up with assumptions, which is not the case.